### PR TITLE
libdlt: truncate the log message if it is too long (#156)

### PR DESCRIPTION
**Pull request by votrungchi**
Change the behavior of dlt_user_log_write_string,
dlt_user_log_write_constant_string and dlt_user_log_write_utf8_string
so that they truncate the incoming message if it is too big
and add "Message truncated, too long" at the end of it.

These function should anyway still return an DLT_RETURN_USER_BUFFER_FULL
error in this case.

Signed-off-by: Nguyen Dinh Thi (RBVH/ENG2) <Thi.NguyenDinh@vn.bosch.com>
Signed-off-by: Vo Trung Chi <chi.votrung@vn.bosch.com>

### CODE REVIEW

**Comment by ssugiura** [comment]
I propose to change the summary since it's not specific to DLT Viewer.

**Comment by votrungchi** [comment]
> 
> 
> I propose to change the summary since it's not specific to DLT Viewer.

Yes, I agree with you. Thank @ssugiura

File: src/lib/dlt_user.c [hunk 1]
```diff
@@ -2333,31 +2305,119 @@ DltReturnValue dlt_user_log_write_constant_string(DltContextData *log, const cha
...
+    uint16_t arg_size = 0;
+    uint32_t type_info = 0;
     size_t new_log_size = 0;
+    DltReturnValue ret = DLT_RETURN_OK;
+
+    size_t str_truncate_message_length = strlen(STR_TRUNCATED_MESSAGE) + 1;
+    size_t max_payload_str_msg;
 
     if ((log == NULL) || (text == NULL))
+    {
```
Comment (drLeo) [inline src/lib/dlt_user.c]: Please be consistent in coding style!

Comment (thanhbnq) [inline src/lib/dlt_user.c]: Please be consistent in coding style!

Comment (votrungchi) [reply src/lib/dlt_user.c]: Fixed.

File: src/lib/dlt_user.c [hunk 2]
```diff
@@ -310,7 +309,8 @@ static DltReturnValue dlt_initialize_fifo_connection(void)
 
     if (ret == -1)
         dlt_vnlog(LOG_WARNING, DLT_USER_BUFFER_LENGTH, "Loging disabled, FIFO user %s cannot be created!\n", filename);
-        /* return DLT_RETURN_OK; */ /* removed to prevent error, when FIFO already exists */
+
+    /* return DLT_RETURN_OK; */ /* removed to prevent error, when FIFO already exists */
```
[the remark this answers is not in the document]
Comment (votrungchi) [reply src/lib/dlt_user.c]: Done.

File: src/lib/dlt_user.c [hunk 3]
```diff
@@ -3593,10 +3577,9 @@ DltReturnValue dlt_user_log_send_log(DltContextData *log, int mtype)
         /* In verbose mode, send extended header */
         msg.standardheader->htyp = (msg.standardheader->htyp | DLT_HTYP_UEH);
     else
-        /* In non-verbose, send extended header if desired */
-        if (dlt_user.use_extende_header_for_non_verbose)
-            msg.standardheader->htyp = (msg.standardheader->htyp | DLT_HTYP_UEH);
-
+    /* In non-verbose, send extended header if desired */
+    if (dlt_user.use_extende_header_for_non_verbose)
```
[the remark this answers is not in the document]
Comment (votrungchi) [reply src/lib/dlt_user.c]: Done.

File: src/lib/dlt_user.c [hunk 4]
```diff
@@ -2299,65 +2333,79 @@ DltReturnValue dlt_user_log_write_string(DltContextData *log, const char *text)
...
+                    reduce_size = 1;
+                /* Is the next to last byte of truncated text is the first byte in multi-byte sequence (utf8 3 bytes) */
+                else if ((tmp[1] & 0xe0) == 0xe0)
+                    reduce_size = 2;
+                /* utf8 4 bytes */
+                else if ((tmp[0] & 0xf0) == 0xf0)
+                    reduce_size = 3;
+            }
+            else {
+                /* one byte utf8, do-nothing */
```
[the remark this answers is not in the document]
Comment (votrungchi) [reply src/lib/dlt_user.c]: Done.

**Pull request merged**